src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -1,21 +1,506 @@
+fillRangeFinderToEnd :: MCompactIndex s k -> ST s ()
+fillRangeFinderToEnd MCompactIndex{..} = do
+    lastMinRfbits <- readSTRef mciLastMinRfbits
+    let lb = smaybe NoBound (BoundExclusive . fromIntegral) lastMinRfbits
+        ub = NoBound
+        x  = fromIntegral $ VUM.length mciPrimary
+    writeRange mciRangeFinder lb ub x
+    writeSTRef mciLastMinRfbits (SJust $ 2 ^ mciRangeFinderPrecision)
+
+{-------------------------------------------------------------------------------
+  Chunks
+-------------------------------------------------------------------------------}
+
+data Chunk k = Chunk {
+    cPrimary :: !(VU.Vector Word32)
+  , cClashes :: !(VU.Vector Bit)
+  }
+
+data FinalChunk k = FinalChunk {
+    fcRangeFinder          :: !(VU.Vector Word32)
+  , fcRangeFinderPrecision :: !Int
+  , fcTieBreaker           :: !(IntMap k)
+  }
+
+fromChunks :: [Chunk k] -> FinalChunk k -> CompactIndex k
+fromChunks cs FinalChunk{..} = CompactIndex {
+      ciRangeFinder          = fcRangeFinder
+    , ciRangeFinderPrecision = fcRangeFinderPrecision
+    , ciPrimary              = VU.concat $ fmap cPrimary cs
+    , ciClashes              = VU.concat $ fmap cClashes cs
+    , ciTieBreaker           = fcTieBreaker
+    }
+
+sliceChunk :: Int -> Int -> MCompactIndex s k -> ST s (Chunk k)
+sliceChunk pageNr n MCompactIndex{..} = do
+    cPrimary <- VU.freeze $ VUM.slice pageNr n mciPrimary
+    cClashes <- VU.freeze $ VUM.slice pageNr n mciClashes
+    pure Chunk{..}
+
+sliceChunkUnsafe :: Int -> Int -> MCompactIndex s k -> ST s (Chunk k)
+sliceChunkUnsafe pageNr n MCompactIndex{..} = do
+    cPrimary <- VU.unsafeFreeze $ VUM.slice pageNr n mciPrimary
+    cClashes <- VU.unsafeFreeze $ VUM.slice pageNr n mciClashes
+    pure Chunk{..}
+
+getFinalChunk :: MCompactIndex s k -> ST s (FinalChunk k)
+getFinalChunk mci@MCompactIndex{..} = do
+    fillRangeFinderToEnd mci
+
+    fcRangeFinder <- VU.freeze mciRangeFinder
+    fcTieBreaker <- readSTRef mciTieBreaker
+    pure FinalChunk{fcRangeFinderPrecision = mciRangeFinderPrecision, ..}
+
+getFinalChunkUnsafe :: MCompactIndex s k -> ST s (FinalChunk k)
+getFinalChunkUnsafe mci@MCompactIndex{..} = do
+    fillRangeFinderToEnd mci
+
+    fcRangeFinder <- VU.unsafeFreeze mciRangeFinder
+    fcTieBreaker <- readSTRef mciTieBreaker
+    pure FinalChunk{fcRangeFinderPrecision = mciRangeFinderPrecision, ..}
+
+{-------------------------------------------------------------------------------
+  SliceBits
+-------------------------------------------------------------------------------}
+
+class SliceBits a where
+  -- | @'sliceBits' offset size x@ slices from @x@ a string of bits of size
+  -- @n@ that starts at the 0-based @offset@.
+  --
+  -- Offsets are counted from the /top/, which means offset @0@ corresponds to
+  -- the most significant bit.
+  sliceBits :: Int -> Int -> a -> a
+
+  -- | Like 'sliceBits', but the offset is always @0@.
+  topBits :: Int -> a -> a
```

Comment:
I would suggest that we change this to have both slice and top bits always return Word32 or Word64 rather than the input type a.

Then we don't need to treat the input type like a numeric type, it just needs to be something we can slice bits out of, such as a byte string key.